File: chapter 1/1.5.py
# 1.5
# One Away: There are three types of edits that can be performed on strings: insert a character,
# remove a character, or replace a character. Given two strings, write a function to check if they are
# one edit (or zero edits) away. 

def OneAway(str1, str2):
    # quick exits
    if str1 == str2:
        return True
    if abs( len(str1) - len(str2) ) > 1:
        return False
    
    # if string lengths are the same
    editCount = 0
    if len(str1) == len(str2):
        for i in range(0, len(str1)):
            if str1[i] != str2[i]:
                editCount +=1
                if editCount > 1:
                    return False
        return True

    # if strings' lengths are different
    editCount = 0
    i = 0
    j = 0
    if len(str1) > len(str2):
        longString = str1
        shortString = str2
    if len(str2) > len(str1):
        longString = str2
        shortString = str1

    while j < len(shortString):
        if longString[i] != shortString[j]:
            editCount += 1
            i += 1
        else:
            i += 1
            j += 1

        if editCount > 1:
            return False

    return True

    # check below if a character is deleted before the end    
    # if shortString[len(shortString)] != longString[len(longString)]

testString0 = "pile"
testString1 = "pale"
testString2 = "pike"
testString3 = "pal"
testString4 = "paler"
testString5 = "pat"
testString6 = "apt"
testString7 = "pie"

print(OneAway(testString0, testString1))

File: chapter 1/js/1.5.js
// One Away: There are three types of edits that can be performed on strings: insert a character,
// remove a character, or replace a character. Given two strings, write a function to check if they are
// one edit (or zero edits) away.
// EXAMPLE
// pale , pi e - > tru e
// pales , pal e - > tru e
// pale , bal e - > tru e
// pale , bak e - > fals e 

function oneAway(str1,str2){
    if (Math.abs(str1.length - str2.length)>1){
        return false;
    }
    if (str1 == str2){
        return true;
    }
    let oneDifferent = false;
    let strL = str1;
    let strS = str2;
    if (str1.length < str2.length){
        strL = str2;
        strS = str1;
    }
    let longStrIndex = 0;
    let shortStrIndex = 0;

    while(shortStrIndex < strS.length){
        console.log(strS[shortStrIndex])
        if(strS[shortStrIndex] != strL[longStrIndex]){
                if (oneDifferent){
                    return false
                }
                oneDifferent = true;
            if(strS.length == strL.length){
                shortStrIndex ++;
            }
        } else {
            shortStrIndex ++;
        }
        longStrIndex ++;
    }
    return true;
}

testString0 = "pile"
testString1 = "pale"
testString2 = "pike"
testString3 = "pal"
testString4 = "paler"
testString5 = "pat"
testString6 = "apt"
testString7 = "pie"

console.log(oneAway(testString7, testString0))